### PR TITLE
Use IProblemDetailService in ValidationProblem

### LINKED ISSUE #47631: Use IProblemDetailsService in ValidationProblem
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Validation problem should use IProblemDetailsService in the same way as #47045 

### Describe the solution you'd like

I think that ValidationProblem (Microsoft.AspNetCore.Http.HttpResults) should also use the IProblemDetailsService in the same way as #47045 to be able to customize the validation problem in one place.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dnperfors**
# Use IProblemDetailService in ValidationProblem

ValidationProblem should use IProblemDetailService so that project wide customizations can be done

## Description

This is basically a modified version of #47100

Fixes #47631

### CODE REVIEW

**Comment by amcasey** [comment]
Seems like a faithful port of the previous PR.  Could/should `JsonHttpResult<ProblemDetails>` as well?

**Comment by amcasey** [comment]
/azp run

(Failures look related to node download issues)

**Comment by dnperfors** [comment]
> Seems like a faithful port of the previous PR.  Could/should `JsonHttpResult<ProblemDetails>` as well?

Good catch, will add that after the weekend.

**Review by captainsafia — APPROVED** [review]
LGTM! I'm OK with factoring out the work to update how `JsonHttpResult<ProblemDetails>` is handled to a different PR.

File: src/Http/Http.Results/src/ValidationProblem.cs [hunk 1]
```diff
@@ -49,21 +49,25 @@ internal ValidationProblem(HttpValidationProblemDetails problemDetails)
...
 
         var loggerFactory = httpContext.RequestServices.GetRequiredService<ILoggerFactory>();
         var logger = loggerFactory.CreateLogger(typeof(ValidationProblem));
+        var problemDetailsService = httpContext.RequestServices.GetService<IProblemDetailsService>();
 
         HttpResultsHelper.Log.WritingResultAsStatusCode(logger, StatusCode);
         httpContext.Response.StatusCode = StatusCode;
 
-        return HttpResultsHelper.WriteResultAsJsonAsync(
+        if (problemDetailsService is null || !await problemDetailsService.TryWriteAsync(new() { HttpContext = httpContext, ProblemDetails = ProblemDetails }))
```
Comment (davidfowl) [inline src/Http/Http.Results/src/ValidationProblem.cs:63]: The performance regression here is a little unfortunate:
- async state machine
- ProblemsDetailsContext object

Is there no way to optimize this?

Comment (amcasey) [reply src/Http/Http.Results/src/ValidationProblem.cs:63]: @davidfowl I don't see any synchronous APIs on `IProblemDetailService`.

I imagine http results are very hot code, but this should be scoped to a particular class of failure results, shouldn't it?

**Comment by bjornen77** [comment]
ping

**Comment by dnperfors** [comment]
I am not sure what is expected here. I didn't implement `JsonHttpResult<ProblemDetails>` because @captainsafia was ok with a separate or for that.
But I don't know what to do about the remark of @davidfowl ...
With some guidance I might be able to resolve it, but I can't think of a solution ...

**Comment by captainsafia** [comment]
@dnperfors Thanks so much for your patience here!

I'm going to go ahead and merge this change. The perf issue that was identified above isn't new to the codebase, we use a similar strategy in `ProblemHttpResult` and I don't want to put the onus on you to fix a pattern you haven't introduced to the codebase. I've filed https://github.com/dotnet/aspnetcore/issues/56439 to track profiling and fixing this.

**Pull request merged**